peps/pep-0747.rst

```diff
@@ -444,25 +444,17 @@ so must be disambiguated based on its argument type:
 the value ``int1 | int2``, ``set1 | set2``, ``dict1 | dict2``, and more,
 so must be disambiguated based on its argument types:
 
--  As a value expression, ``x | y`` has type equal to the return type of ``type(x).__or__``
-   if ``type(x)`` overrides the ``__or__`` method.
-
-   -  When ``x`` has type ``builtins.type``, ``types.GenericAlias``, or the
-      internal type of a typing special form, ``type(x).__or__`` has a return type
-      in the format ``TypeExpr[T1 | T2]``.
-
--  As a value expression, ``x | y`` has type equal to the return type of ``type(y).__ror__``
-   if ``type(y)`` overrides the ``__ror__`` method.
-
-   -  When ``y`` has type ``builtins.type``, ``types.GenericAlias``, or the
-      internal type of a typing special form, ``type(y).__ror__`` has a return type
-      in the format ``TypeExpr[T1 | T2]``.
-
--  As a value expression, ``x | y`` has type ``UnionType``
-   in all other situations.
-
-   -  This rule is intended to be consistent with the preexisting fallback rule
-      used by static type checkers.
+-  As a value expression, ``x | y`` was given type ``UnionType`` in certain situations
```

Comment:
The rules for type evaluation of unions in value expressions were previously unspecified. The PEP needs to specify the intended behavior, and this explanation cannot depend on a past unspecified behavior. The way this is currently worded, it's not at all clear what "certain circumstances" means. It needs to be clear enough that we can write a definitive conformance test for the behavior.

Keep in mind that `UnionType` does not appear in any typing PEPs or in the typing spec. It is not part of the static type system, and it has no defined meaning in this context. I think it's fine to refer to `UnionType` in the "Backward compatibility" section of the PEP, but it shouldn't appear in the "Specification" section. The "Specification" section should be written such that it can be incorporated directly into the typing spec once the PEP is accepted.

Reply:
> The rules for type evaluation of unions in value expressions were previously unspecified. The PEP needs to specify the intended behavior, and this explanation cannot depend on a past unspecified behavior. The way this is currently worded, it's not at all clear what "certain circumstances" means.

@erictraut the new proposed wording intentionally leans on previous type checker behavior, albeit unspecified and perhaps varying across checkers, and attempts to apply a minimal diff to it. I do not personally have the energy/time available to attempt to deduce the current unspecified behavior with many back and forth conversations or by reading the code of multiple checkers.

The remaining time I have available to dedicate to this PEP is rapidly running out. Therefore I think I will sidestep all of the `X | Y` specification work by simply requiring users to write `TypeExpr(X | Y)` explicitly if they want to use a union type in a TypeExpr context. I'll make a new PR that takes this alternative approach.

Somebody else can clear up the `X | Y` rules in the future.

Reply:
@davidfstr, thanks for taking the PEP to this stage. I appreciate all the work you've put into it. When you're ready to "hand the pen to someone else", I'm willing to do a revised draft that addresses this issue and others that I still see as being problematic in latest draft. I'm hopeful that we can get this over the finish line.